### PR TITLE
[mlir][tensor] Fold pack and unpack of empty input tensor

### PR DESCRIPTION
**Pull request by adam-smnk**
Extends `tensor.empty` folding patterns with pack and unpack consumers to fold away the operations when their source is empty.

### CODE REVIEW

**Comment by adam-smnk** [comment]
It also could be an optional pattern maybe in `populateSimplifyPackAndUnpackPatterns` or `populateFoldTensorEmptyPatterns`.

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 1]
```diff
@@ -4200,6 +4200,12 @@ LogicalResult PackOp::canonicalize(PackOp packOp, PatternRewriter &rewriter) {
     return success();
   }
 
+  // Fold away packing an empty source tensor.
+  if (auto emptyTensor = packOp.getSource().getDefiningOp<tensor::EmptyOp>()) {
+    rewriter.replaceOp(packOp, packOp.getDest());
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:4205]: I think this could be done in the folder.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Thanks for adding these. I don't think these should be folders or canonicalizations. There are patterns that do folding of tensor.empty and it's consumers, these are better suited there.

Also packs have padding semantics. Those need to be accounted for. I think if there is padding you can't do this transformation

**Comment by adam-smnk** [comment]
Thanks for the feedback. I'll move it to the folder patterns.

> Also packs have padding semantics. Those need to be accounted for. I think if there is padding you can't do this transformation

At first I treated packing on empty as mostly UB but I suppose you could materialize it as a valid initialization of a buffer.
I'll make the transform more conservative.

**Comment by MaheshRavishankar** [comment]
> Thanks for the feedback. I'll move it to the folder patterns.
> 
> > Also packs have padding semantics. Those need to be accounted for. I think if there is padding you can't do this transformation
> 
> At first I treated packing on empty as mostly UB but I suppose you could materialize it as a valid initialization of a buffer. I'll make the transform more conservative.

This is where you should add this https://github.com/shark-infra/llvm-project/blob/f44eaa13d0c56630791e6484fa3d86a813fa716c/mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp#L98

File: mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp [hunk 2]
```diff
@@ -93,12 +93,76 @@ struct FoldEmptyTensorWithExtractSliceOp
...
+
+  LogicalResult matchAndRewrite(PackOp packOp,
+                                PatternRewriter &rewriter) const override {
+    // Check for tensor.empty source.
+    auto emptyOp = packOp.getSource().getDefiningOp<EmptyOp>();
+    if (!emptyOp)
+      return failure();
+
+    // Check for single use.
+    if (foldSingleUseOnly && !llvm::hasSingleElement(emptyOp->getUses()))
```
Comment (chelini) [inline mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp:112]: You can simply write: !emptyOp->hasOneUse()

**Review by chelini — COMMENTED** [review]
Thanks for the ping @hanhanW.

File: mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp [hunk 3]
```diff
@@ -93,12 +93,76 @@ struct FoldEmptyTensorWithExtractSliceOp
   bool foldSingleUseOnly = false;
 };
 
+/// tensor.empty does not define any tensor contents, so an unpadded pack
```
[the remark this answers is not in the document]
Comment (adam-smnk) [reply mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp:96]: Padding has a side effect - if it kicks in, you get at least partially initialized buffer.
So, I opted for a more conservative approach. You could probably replace pack with pad or sth but leaving it for the future.

File: mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (adam-smnk) [reply mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp:112]: Just following this patterns' API.

Comment (adam-smnk) [reply mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp:112]: Agreed but here I just copy-pasted for consistency with the rest of the file.

File: mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (chelini) [reply mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp:96]: To be fair, I don't quite get. How is it possible for a tensor.empty (or a tensor) to be partially initialed? Do you have an example I can take a look at?

[the remark this answers is not in the document]
Comment (adam-smnk) [reply mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp:96]: Sure, for example:
```
func.func @entry(%arg0: tensor<1x1x32x32xf32>) -> tensor<1x1x32x32xf32> {
  %pad = arith.constant 3.0 : f32
  %empty_unpacked = tensor.empty() : tensor<20x20xf32>
  %packed = tensor.pack %empty_unpacked padding_value(%pad : f32)
    inner_dims_pos = [0, 1] inner_tiles = [32, 32]
    into %arg0 : tensor<20x20xf32> -> tensor<1x1x32x32xf32>
  return %packed : tensor<1x1x32x32xf32>
}
```

Assuming `%empty_unpacked = tensor.empty()` gets lowering into `memref.alloc()`, the original shape `20x20` will contain random uninitialized data. However, the remaining padding up to `32x32` will have the predefined `%pad` value (here 3.0).

File: mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp:112]: I dont think you need this. You can drop it I think. I dont know why the other patterns do that, probably being too conservative,

File: mlir/test/Dialect/Tensor/fold-empty-op.mlir [hunk 4]
```diff
@@ -64,6 +64,47 @@ func.func @rank_reducing_empty_tensor_extract(%sz : index, %idx : index) -> tens
   return %r: tensor<2xf32>
 }
 
+func.func @pack_empty(%arg0: tensor<8x8x32x32xf32>) -> tensor<8x8x32x32xf32> {
```
Comment (MaheshRavishankar) [inline mlir/test/Dialect/Tensor/fold-empty-op.mlir:67]: Could you add a dynamic shape test?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Thanks! Just a small comment.

File: mlir/test/Dialect/Tensor/fold-empty-op.mlir (hunk 4 above)
Comment (adam-smnk) [reply mlir/test/Dialect/Tensor/fold-empty-op.mlir:67]: Added

File: mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (adam-smnk) [reply mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp]: Dropped the check for pack/unpack.
I don't think there's any scenario where multiple tensor empty users could prevent this fold pattern.

File: mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (chelini) [reply mlir/lib/Dialect/Tensor/Transforms/EmptyOpPatterns.cpp:96]: right, I see. I think in this case we can "strength reduce" to pad but not in this PR.

File: mlir/test/Dialect/Tensor/fold-empty-op.mlir [hunk 5]
```diff
@@ -64,6 +64,73 @@ func.func @rank_reducing_empty_tensor_extract(%sz : index, %idx : index) -> tens
...
+func.func @pack_empty_dynamic(%arg0: tensor<?x?x?x?xf32>, %dim0: index, %dim1: index) -> tensor<?x?x?x?xf32> {
+  %empty_unpacked = tensor.empty(%dim0, %dim1) : tensor<?x?xf32>
+  %packed = tensor.pack %empty_unpacked
+    inner_dims_pos = [0, 1] inner_tiles = [32, 32]
+    into %arg0 : tensor<?x?xf32> -> tensor<?x?x?x?xf32>
+  return %packed : tensor<?x?x?x?xf32>
+}
+
+// CHECK-LABEL: func.func @pack_empty_dynamic(
+// CHECK-SAME:   %[[T:.+]]: tensor<?x?x?x?xf32>
```
Comment (MaheshRavishankar) [inline mlir/test/Dialect/Tensor/fold-empty-op.mlir:89]: Nit: Please check that the tensor shape is what you expect by looking at the operands for `tensor.empty`

File: mlir/test/Dialect/Tensor/fold-empty-op.mlir [hunk 6]
```diff
@@ -64,6 +64,73 @@ func.func @rank_reducing_empty_tensor_extract(%sz : index, %idx : index) -> tens
...
+func.func @unpack_empty_dynamic(%arg0: tensor<?x?xf32>, %dim0: index, %dim1: index, %dim2: index, %dim3: index) -> tensor<?x?xf32> {
+  %empty_packed = tensor.empty(%dim0, %dim1, %dim2, %dim3) : tensor<?x?x?x?xf32>
+  %unpacked = tensor.unpack %empty_packed
+    inner_dims_pos = [0, 1] inner_tiles = [32, 32]
+    into %arg0 : tensor<?x?x?x?xf32> -> tensor<?x?xf32>
+  return %unpacked : tensor<?x?xf32>
+}
+
+// CHECK-LABEL: func.func @unpack_empty_dynamic(
+// CHECK-SAME:   %[[T:.+]]: tensor<?x?xf32>
```
Comment (MaheshRavishankar) [inline mlir/test/Dialect/Tensor/fold-empty-op.mlir:115]: Same here. Check that the `tensor.empty` operands are what you expect.

**Review by MaheshRavishankar — APPROVED** [review]
Have a few comments on the lit tests. Looks good otherwise. Please address before landing.

**Pull request merged**